src/main/scala/org/broadinstitute/dsde/workbench/sam/service/ResourceService.scala

```diff
@@ -302,15 +322,16 @@ class ResourceService(
         case Some(error) => IO.raiseError(new WorkbenchExceptionWithErrorReport(error.copy(statusCode = Option(StatusCodes.BadRequest))))
         case None =>
           val newMembers = emailsToSubjects.values.flatten.toSet
-          accessPolicyDAO.loadPolicy(policyId, samRequestContext).flatMap {
-            case None => IO.raiseError(new WorkbenchExceptionWithErrorReport(ErrorReport(StatusCodes.NotFound, s"policy $policyId does not exist")))
-            case Some(existingPolicy) if existingPolicy.members == newMembers => IO.unit
-            case _ => for {
-              _ <- accessPolicyDAO.overwritePolicyMembers(policyId, newMembers, samRequestContext)
-              _ <- fireGroupUpdateNotification(policyId, samRequestContext)
-            } yield ()
+          accessPolicyDAO.listAccessPolicies(policyId.resource, samRequestContext).flatMap { originalPolicies =>
+            originalPolicies.find(_.id == policyId) match {
+              case None => IO.raiseError(new WorkbenchExceptionWithErrorReport(ErrorReport(StatusCodes.NotFound, s"policy $policyId does not exist")))
+              case Some(existingPolicy) if existingPolicy.members == newMembers => IO.unit
+              case _ => for {
+                _ <- accessPolicyDAO.overwritePolicyMembers(policyId, newMembers, samRequestContext)
+                _ <- onPolicyUpdate(policyId, originalPolicies, samRequestContext)
+              } yield ()
```

Comment:
I got a bit confused by more than one case in a pattern match on `Option` because it can only take two forms. This is where I think applicative syntax really adds value. Your choice, naturally.
```suggestion
              case Some(existingPolicy) => Applicative[IO].whenA(existingPolicy.members != newMembers) {
                for {
                  _ <- accessPolicyDAO.overwritePolicyMembers(policyId, newMembers, samRequestContext)
                  _ <- onPolicyUpdate(policyId, originalPolicies, samRequestContext)
                } yield ()
              }
```